src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -160,46 +254,512 @@ private static bool TryGetCompatibleWindowsRuntimeTypesForVariantType(INamedType
+                    if (genericStartIdx != -1)
+                    {
+                        interfaceStaticsMethod += @interface[genericStartIdx..@interface.Length];
+                    }
+
+                    source.AppendLine($$"""
+                                                new global::System.Runtime.InteropServices.ComWrappers.ComInterfaceEntry
+                                                {
+                                                    IID = global::ABI.{{interfaceStaticsMethod}}.IID,
+                                                    Vtable = global::ABI.{{interfaceStaticsMethod}}.AbiToProjectionVftablePtr
+                                                },
+                                    """);
+                }
+                source.AppendLine($$"""
+                                };
+                                """);
+            }
+            else
+            {
+                source.AppendLine($$"""
+                                        return global::System.Array.Empty<global::System.Runtime.InteropServices.ComWrappers.ComInterfaceEntry>();
+                                """);
+            }
+
+            return source.ToString();
+        }
+
+        internal static void GenerateVtableAttributes(Action<string, string> addSource, ImmutableArray<VtableAttribute> vtableAttributes)
+        {
+            // Using ToImmutableHashSet to avoid duplicate entries from the use of partial classes by the developer
+            // to split out their implementation.  When they do that, we will get multiple entries here for that
+            // and try to generate the same attribute and file with the same data as we use the semantic model
+            // to get all the symbol data rather than the data at an instance of a partial class definition.
+            foreach (var vtableAttribute in vtableAttributes.ToImmutableHashSet())
+            {
+                if (vtableAttribute.Interfaces.Any())
+                {
+                    StringBuilder source = new();
+                    source.AppendLine("using static WinRT.TypeExtensions;\n");
+                    if (!vtableAttribute.IsGlobalNamespace)
+                    {
+                        source.AppendLine($$"""
                         namespace {{vtableAttribute.Namespace}}
                         {
                         """);
-                }
+                    }
 
-                source.Append($$"""
-                    [global::WinRT.WinRTExposedType({{vtableEntries}})]
+                    source.AppendLine($$"""
+                    [global::WinRT.WinRTExposedType(typeof({{vtableAttribute.ClassName}}WinRTTypeDetails))]
                     partial class {{vtableAttribute.ClassName}}
                     {
                     }
+
+                    internal sealed class {{vtableAttribute.ClassName}}WinRTTypeDetails : global::WinRT.IWinRTExposedTypeDetails

```

Comment:
Question: right now we have this `WinRTExposedType`, which links to a separate generated type with the details, implementing the interface, which is constructed at runtime using reflection. Could we not do the same as `ComWrappers` is doing instead, that is, make `WinRTExposedType` non-sealed, and then have the generator create a derived type directly implementing the interface, and use that to annotate the projection types? This way CsWinRT can simply look that up and cast directly to the interface type, no need to use `Activator.CreateInstance` and reflection to get an instance to invoke methods on. Basically it'd just generate this:
```csharp
[FooWinRTExposedType]
partial class Foo
{
}

internal sealed class FooWinRTExposedType : WinRTExposedType, IWinRTExposedTypeDetails
{
    public ComWrappers.ComInterfaceEntry[] GetExposedInterfaces() => ...;
}
```
Then CsWinRT can just do `(IWinRTExposedTypeDetails?)type.GetCustomAttribute(typeof(WinRTExposedType))` and use that directly. Should also help startup a tiny bit since there's no more `Activator.CreateInstance` being called. And this also works downlevel just fine, as we don't even need any attribute generics.

What do you think? 🙂

Reply:
This was one of my initial ideas during the refactor.  This might still be preferrable to do, will need to do some comparisons.  The main reason I switched to what I have was I was concerned about generating much more code for each struct, especially blittable ones, where we only generate the struct today and not any marshaling classes.  With the current approach, I was able to use a generic type that I passed to share the implementation for structs.  But after this PR is in, I am willing to try out the other approach again and see how much more it really adds to code size and if it is minimal.  Also I was thinking this approach may make it easier in the future to move to generic attributes when we can, but with that said, not sure which is better perf.